programs/locker/src/instructions/v3/claim.rs

```diff
@@ -0,0 +1,204 @@
+use anchor_spl::token_interface::{Mint, TokenAccount, TokenInterface};
+use safe_math::SafeMath;
+use solana_program::hash::hashv;
+use util::{
+    parse_remaining_accounts, AccountsType, ParsedRemainingAccounts, TRANSFER_MEMO_CLAIM_VESTING,
+};
+
+const LEAF_PREFIX: &[u8] = &[0];
+const INTERMEDIATE_PREFIX: &[u8] = &[1];
+
+#[derive(AnchorSerialize, AnchorDeserialize)]
+pub struct ClaimV3Params {
+    pub vesting_start_time: u64,
+    pub cliff_time: u64,
+    pub frequency: u64,
+    pub cliff_unlock_amount: u64,
+    pub amount_per_period: u64,
+    pub number_of_period: u64,
+    pub max_amount: u64,
+    pub proof: Vec<[u8; 32]>,
+}
+
+impl ClaimV3Params {
+    pub fn verify_recipient(&self, recipient: Pubkey, root: [u8; 32]) -> Result<()> {
+        let node = hashv(&[
+            &recipient.key().to_bytes(),
+            &self.cliff_unlock_amount.to_le_bytes(),
+            &self.amount_per_period.to_le_bytes(),
+            &self.number_of_period.to_le_bytes(),
+            &self.cliff_time.to_le_bytes(),
+            &self.frequency.to_le_bytes(),
+            &self.vesting_start_time.to_le_bytes()
+        ]);
+
+        let leaf = hashv(&[LEAF_PREFIX, &node.to_bytes()]);
+
+        let mut computed_hash = leaf.to_bytes();
+        for p in self.proof.iter() {
+            if computed_hash <= *p {
+                computed_hash = hashv(&[&INTERMEDIATE_PREFIX, &computed_hash, p]).to_bytes();
+            } else {
+                computed_hash = hashv(&[&INTERMEDIATE_PREFIX, p, &computed_hash]).to_bytes();
+            }
+        }
+
+        require!(computed_hash == root, LockerError::InvalidMerkleProof);
+
+        Ok(())
+    }
+
+    pub fn get_max_unlocked_amount(&self, current_ts: u64) -> Result<u64> {
+        if current_ts < self.cliff_time {
+            return Ok(0);
+        }
+        let period = current_ts
+            .safe_sub(self.cliff_time)?
+            .safe_div(self.frequency)?;
+        let period = period.min(self.number_of_period);
+
+        let unlocked_amount = self
+            .cliff_unlock_amount
+            .safe_add(period.safe_mul(self.amount_per_period)?)?;
+
+        Ok(unlocked_amount)
+    }
+
+    pub fn get_claimable_amount(&self, total_claimed_amount: u64, current_ts: u64) -> Result<u64> {
+        let max_unlocked_amount = self.get_max_unlocked_amount(current_ts)?;
+        let claimable_amount = max_unlocked_amount.safe_sub(total_claimed_amount)?;
+        Ok(claimable_amount)
+    }
+
+    pub fn get_claim_amount(&self, total_claimed_amount: u64) -> Result<u64> {
+        let current_ts = Clock::get()?.unix_timestamp as u64;
+        let claimable_amount = self.get_claimable_amount(total_claimed_amount, current_ts)?;
+
+        let amount = claimable_amount.min(self.max_amount);
+
+        Ok(amount)
+    }
+}
+
+/// Accounts for [locker::claim_v3].
+#[event_cpi]
+#[derive(Accounts)]
+pub struct ClaimV3<'info> {
+    /// Claim status PDA
+    #[account(
+        init,
```

Comment:
why use `init`? It meant each recipient can only call this function 1 time, so they can't claim all tokens

Reply:
i updated it.